designs/limits.md

```diff
@@ -0,0 +1,60 @@
+
+
+## Proposed Solution for Limits
+
+There are two broad forms of limiting we could apply. The first is that we could introduce a limit to the number of in-flight worker node being provisioned at a point in time. A worker node that's in the `NotReady` state could be considered to be in-flight. The second form is an absolute limit of the number of resources Karpenter can provision.
+
+
+### **In-flight limit**
+
+```yaml
+spec:
+  limits:
+    unready: 20% # StringOrInt
+```
+
+The good bit about this approach is that we don't artificially constrain how many total worker nodes can be spun up by Karpenter, while also making sure that if we're launching worker nodes that aren't healthy we stop the scaling and save costs.
+
+The two main problems with this approach though are -
+1. This limit while meant to constrain the number of unhealthy worker nodes in a cluster, will also inihibit the rate at which Karpenter can respond to pods that aren't schedulable. This somewhat goes against the goal of minimizing launch times of workers.
+2. While this helps ensure that costs don't increase due to runaway scaling, it won't help those who want a stricter cap on the amount of resources that's being provisioned even when nodes are otherwise healthy.
+
+### **Absolute limit**
+
+This is a much simpler semantic to follow, where you provide an upper cap on the amount of resources being provisioned irrespective of their health. This is more in-line with alternative solutions like CA and should be easier to reason with.
+
+The actual limit of the number of resources being spun up can be defined either through something similar to resource requests or just a flat count of instances.
+
+```yaml
+spec:
+  limits:
+    resources:
+      cpu: 1000
+      memory: 1000Gi
+```
+
+Modeling your limit via CPU and Memory constraints aligns well with customers that use resource requests with their pods. This should give everyone a consistent interface to define the capacity requirements of their applications as well as their underlying infrastructure. There shouldn't be a need to think of the actual number and type of instances that are being chosen by Karpenter.
+
+As a cost control mechanism, this requires a little more work from our users if they're looking for higher precision. Since some cloud providers like AWS charge differently per instance type rather than the actual resource request and consumption, a little more translation will be needed to estimate costs. The alternative of a flat count of instance types - say `spec.limits.resources.count`, doesn't help solve that problem either since the instance types are chosen by Karpenter by default. In case overrides are defined via `kubernetes.io/instance-type`, you should be able to calculate the CPU and Memory bounds fairly trivially through some cloud provider API calls.
```

Comment:
IMO this is a good enough limits mechanism for the specific case. I don't think we really want this to be a general cost control mechanism, but rather a "something is very wrong, let's stop" type of control.

A general cost mechanism would be difficult w/ all the purchase types for an individual instance type, like spot vs on-demand, reserved instances, savings plans, etc. 

I'm just agreeing with your point that it's hard and not worth calculating at a higher precision.